lib/Controller/ExternalApiController.php

```diff
@@ -0,0 +1,158 @@
+        $this->groupManager = $groupManager;
+        $this->logger = $logger;
+    }
+
+   public function index($operation)
+   {
+      switch($operation) {
+         case 'auth':
+            return checkPassword(
+               $this->request->getParam('username'),
+               $this->request->getParam('password'),
+               $this->request->getParam('domain')
+            );
+            break;
+         case 'isuser':
+            return isUser(
+               $this->request->getParam('username'),
+               $this->request->getParam('domain')
+            );
+            break;
+         case 'sharedroster':
+            return sharedRoster(
+               $this->request->getParam('username'),
+               $this->request->getParam('domain')
+            );
+            break;
+         default:
+            throw new UnprocessableException( 'Unsupported operation.');
+      }
+   }
+
+   public function checkPassword($username = '', $password = '', $domain = '') {
+      $currentUser = null;
+
+      $this->logger->info('ExAPI: Check password for user: '.$username.'@'.$domain);
+
+      if(!empty($password) && !empty($username)) {
+         if(!empty($domain)) {
+            $loggedIn = $this->userSession->login($username . '@' . $domain, $password);
+         }
+         if(!$loggedIn) {
+            $loggedIn = $this->userSession->login($username, $password);
+         }
+
+         if($loggedIn === true) {
+            $currentUser = $this->userSession->getUser();
+         }
+      }
+
+      if (!$currentUser) {
+          return array(
+             'result' => 'noauth',
+          );
+      }
+
+      $data = array();
+      $data ['uid'] = $currentUser->getUID();
+
+      return array(
+         'result' => 'success',
+         'data' => $data,
+      );
+   }
+
+   public function isUser($username = '', $domain = '') {
+      $this->logger->info('ExAPI: Check if "'.$username.'@'.$domain.'" exists');
+
+      $isUser = false;
+
+      if(!empty($username)) {
+         if(!empty($domain)) {
+            $isUser = $this->userManager->userExists($username . '@' . $domain);
+         }
+         if(!$isUser) {
+            $isUser = $this->userManager->userExists($username);
+         }
+      }
+
+      return array(
+         'result' => 'success',
+         'data' => array(
+            'isUser' => $isUser
+         )
+      );
+   }
+
+   public function sharedRoster($username = '', $domain = '') {
+      if(!empty($username)) {
+         if(!empty($domain)) {
+            $username .= '@' . $domain;
+         }
+      } else {
+         throw new UnprocessableException('No username provided');
+      }
+
+      $roster = [];
+      $user = $this->userManager->get($username);
+
+      $userGroups = $this->groupManager->getUserGroups($user);
+
+      foreach($userGroups as $userGroup) {
+         foreach($userGroup->getUsers() as $user) {
+            $uid = $user->getUID();
+
+            if(!$roster[$uid]) {
```

Comment:
It's probably better to use http://php.net/manual/en/function.array-key-exists.php, this will give a warning https://travis-ci.org/nextcloud/jsxc.nextcloud/jobs/256943398#L377